fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaDataBuilder.java

```diff
@@ -484,111 +577,127 @@ private static void validateRecords(@Nonnull Descriptors.FileDescriptor fileDesc
+            }
+
+            // All union fields must be RECORD (i.e., they cannot be NESTED/UNIONs). The same rule applies to imported union fields too.
+            RecordMetaDataOptionsProto.RecordTypeOptions recordTypeOptions = descriptor.getOptions()
+                    .getExtension(RecordMetaDataOptionsProto.record);
+            if (recordTypeOptions != null &&
+                    recordTypeOptions.hasUsage() &&
+                    recordTypeOptions.getUsage() != RecordMetaDataOptionsProto.RecordTypeOptions.Usage.RECORD) {
+                throw new MetaDataException("Union field " + unionField.getName() +
+                                            " has type " + descriptor.getName() +
+                                            " which is not a record");
+            }
+        }
+
+        // All RECORD message types defined in this proto must be present in the union.
+        for (Descriptors.Descriptor descriptor : fileDescriptor.getMessageTypes()) {
             RecordMetaDataOptionsProto.RecordTypeOptions recordTypeOptions = descriptor.getOptions()
                     .getExtension(RecordMetaDataOptionsProto.record);
-            if (recordTypeOptions != null) {
+            if (recordTypeOptions != null && recordTypeOptions.hasUsage()) {
                 switch (recordTypeOptions.getUsage()) {
-                    case UNION:
-                        if (union != null) {
-                            throw new MetaDataException("Only one union descriptor is allowed");
+                    case RECORD:
+                        if (DEFAULT_UNION_NAME.equals(descriptor.getName())) {
+                            if (unionHasMessageType(unionDescriptor, descriptor)) {
+                                throw new MetaDataException("Union message type " + descriptor.getName()
+                                                            + " cannot be a union field.");
+                            }
+                        } else if (!unionHasMessageType(unionDescriptor, descriptor)) {
+                            throw new MetaDataException("Record message type " + descriptor.getName()
+                                                        + " must be a union field.");
                         }
-                        union = descriptor;
-                        continue;
+                        break;
+                    case UNION:
+                        // Already checked above that none of the union fields is UNION.
                     case NESTED:
-                        continue;
-                    case RECORD:
+                        // Already checked above that none of the union fields is NESTED.
+                    case UNSET:
+                        // Unset message types can appear in either union fields or as nested messages. The only possible issue
+                        // that is the message type's name that might equal "RecordTypeUnion", is already checked above.
```

Comment:
```suggestion
                        // is that the message type's name that might equal "RecordTypeUnion", is already checked above.
```

Sorry, this was my bad (I messed up the order of a few words in a previous suggestion).

Reply:
np.